crates/stack/src/metrics.rs

```diff
@@ -0,0 +1,240 @@
+use std::ops::{Add, Mul, Sub};
+use std::time::{Duration, Instant};
+
+use num_traits::{One, Zero};
+
+pub const DEFAULT_EWMA_ALPHA_SHORT: f32 = 0.9;
+pub const DEFAULT_EWMA_ALPHA_MID: f32 = 0.7;
+pub const DEFAULT_EWMA_ALPHA_LONG: f32 = 0.45;
+
+#[derive(Clone, Default)]
+pub struct ChannelMetrics {
+    pub tx: Metrics,
+    pub rx: Metrics,
+}
+
+#[derive(Clone)]
+pub struct Metrics {
+    pub short: TimeWindow,
+    pub mid: TimeWindow,
+    pub long: TimeWindow,
+}
+
+impl Default for Metrics {
+    fn default() -> Self {
+        Self::new(
+            DEFAULT_EWMA_ALPHA_SHORT,
+            DEFAULT_EWMA_ALPHA_MID,
+            DEFAULT_EWMA_ALPHA_LONG,
+        )
+        .unwrap()
+    }
+}
+
+impl Metrics {
+    pub fn new(short_alpha: f32, mid_alpha: f32, long_alpha: f32) -> Result<Self, f32> {
+        let now = Instant::now();
+
+        Ok(Self {
+            short: TimeWindow::new(now, Duration::from_millis(500), Ewma::new(short_alpha)?),
+            mid: TimeWindow::new(now, Duration::from_secs(1), Ewma::new(mid_alpha)?),
+            long: TimeWindow::new(now, Duration::from_secs(1), Ewma::new(long_alpha)?),
+        })
+    }
+
+    pub fn push(&mut self, value: f32) {
+        let now = Instant::now();
+        self.short.push(value, now);
+        self.mid.push(value, now);
+        self.long.push(value, now);
+    }
+}
+
+/// Series average abstraction
+pub trait Average<T>: Clone {
+    fn push(&mut self, value: T);
+    fn value(&self) -> T;
+}
+
+/// Exponentially weighted moving average
+#[derive(Clone)]
+pub struct Ewma<T = f32>
+where
+    T: Clone,
+{
+    value: Option<T>,
+    alpha: T,
+    one_min_alpha: T,
+}
+
+impl<T> Ewma<T>
+where
+    T: Zero + One + Sub<Output = T> + PartialOrd + Clone,
+{
+    pub fn new(alpha: T) -> Result<Self, T> {
+        let zero = T::zero();
+        let one = T::one();
+
+        if alpha < zero || alpha > one {
+            return Err(alpha);
+        }
+
+        let one_min_alpha = one.sub(alpha.clone());
+
+        Ok(Self {
+            value: None,
+            alpha,
+            one_min_alpha,
+        })
+    }
+}
+
+impl<T> Average<T> for Ewma<T>
+where
+    T: Zero + Add<Output = T> + Mul<Output = T> + Clone,
+{
+    fn push(&mut self, value: T) {
+        let new_value = match self.value.take() {
+            Some(v) => self.alpha.clone().mul(value) + self.one_min_alpha.clone().mul(v),
+            None => value,
+        };
+        self.value.replace(new_value);
```

Comment:
There is no time dependency here. I can't see how average value can be greater than maximal packet size

Reply:
You are right - it seems I broke this code when refactoring the TimeWindow

Reply:
I’ll provide a fix, soon